calibrate.py

```diff
@@ -0,0 +1,438 @@
+            time.sleep(0.1)
+            self.update_display()
+        if button['r1']:
+            self.processor.finished = True
+        if button['r2']:
+            if self.processor.mode_number == 0: self.update_indexs()
+            if self.processor.mode_number == 2: self.update_value()
+            self.remove_controls()
+            self.remove_radio_buttons()
+            self.processor.mode_number = 3
+            self.logger.info("Entering thresholding mode")
+            pygame.mouse.set_visible(True)
+        if button['l1']:
+            if self.processor.mode_number == 0: self.update_indexs()
+            self.remove_radio_buttons()
+            self.processor.mode_number = 2
+            self.logger.info("Manual calibration mode")
+            self.display_controls()
+            pygame.mouse.set_visible(False)
+        if button['l2']:
+            if self.processor.mode_number == 0: self.update_indexs()
+            self.remove_controls()
+            self.remove_radio_buttons()
+            self.processor.mode_number = 1
+            self.logger.info("Auto calibrating mode")
+            pygame.mouse.set_visible(False)
+        #if left or right buttons on right side of joystick pressed, treat them like arrow buttons
+        #if left D pad pressed, treat like tab, double tab or shift tab (tab backwards)
+        if button['circle']:
+            pygame.event.post(pygame.event.Event(pygame.KEYDOWN,{
+                'mod': 0, 'scancode': 77, 'key': pygame.K_RIGHT, 'unicode': "u'\t'"}))
+        if button['square']:
+            pygame.event.post(pygame.event.Event(pygame.KEYDOWN,{
+                'mod': 0, 'scancode': 75, 'key': pygame.K_LEFT, 'unicode': "u'\t'"}))
+        if button['dright']:
+            pygame.event.post(pygame.event.Event(pygame.KEYDOWN,{
+                'mod': 0, 'scancode': 15, 'key': pygame.K_TAB, 'unicode': "u'\t'"}))
+        if button['dleft']:
+            pygame.event.post(pygame.event.Event(pygame.KEYDOWN,{
+                'mod': 1, 'scancode': 15, 'key': pygame.K_TAB, 'unicode': "u'\t'"}))
+        if button['ddown']:
+            pygame.event.post(pygame.event.Event(pygame.KEYDOWN,{
+                'mod': 0, 'scancode': 15, 'key': pygame.K_TAB, 'unicode': "u'\t'"}))
+            pygame.event.post(pygame.event.Event(pygame.KEYDOWN,{
+                'mod': 0, 'scancode': 15, 'key': pygame.K_TAB, 'unicode': "u'\t'"}))
+        if button['dup']:
+            pygame.event.post(pygame.event.Event(pygame.KEYDOWN,{
+                'mod': 1, 'scancode': 15, 'key': pygame.K_TAB, 'unicode': "u'\t'"}))
+            pygame.event.post(pygame.event.Event(pygame.KEYDOWN,{
+                'mod': 1, 'scancode': 15, 'key': pygame.K_TAB, 'unicode': "u'\t'"}))
+
+    def update_value(self):
+        for ctrl in self.controls:
+            i = ctrl['index']
+            self.processor.colour_value[i % 2][int(i/2)] = ctrl['ctrl'].value
```

Comment:
you're calculating this two level accessor for `self.processor.colour_value` 3 times, but it doesn't change. Instead of just getting `i` above, why not do it once, and then use that reference?
```
i = ctrl['index']
colour_val = self.processor.colour_value[i % 2][int(i/2)]
colour_val = ctrl['ctrl'].value
colour_val = int(colour_val)
```

actually, why are we double assigning at all? can't you just:
```
i = ctrl['index']
colour_val = self.processor.colour_value[i % 2][int(i/2)]
colour_val = int(ctrl['ctrl'].value)
```
?

Reply:
I don't think your examples do the thing that's needed, unless colour_val is a pointer. `self.processor.colour_value` is in the form `((low1, low2, low3),(upp1, upp2, upp3))` and needs updating with the control values. the double assigning was weird. `self.processor.colour_value[i % 2][int(i/2)] = int(ctrl['ctrl'].value)` didn't work yet ````self.processor.colour_value[i % 2][int(i/2)] = ctrl['ctrl'].value
self.processor.colour_value[i % 2][int(i/2)] = int(self.processor.colour_value[i % 2][int(i/2)] = ctrl['ctrl'].value)```did work. I guess its something to do with tuples vs lists.  I can try rearranging your logic to work